osrf_testing_tools_cpp/src/memory_tools/CMakeLists.txt

```diff
@@ -48,15 +48,26 @@ if(BFD_LIBRARY)
   target_compile_definitions(memory_tools_interpose PRIVATE BACKWARD_HAS_BFD=1)
 endif()
 
+# Define option to add address sanitizer path to LD_PRELOAD before adding memory tools to the same
+option(OPTIONAL_ASAN_LD_PRELOAD_OVERRIDE "Adds asan lib as a prefix to LD_PRELOAD environment variable" OFF)
+
 # TODO(wjwwood): If on aarch64, so not set the preload environment variables, until fixed.
 # see: https://github.com/osrf/osrf_testing_tools_cpp/issues/3
 if(NOT CMAKE_SYSTEM_PROCESSOR STREQUAL "aarch64")
   if(APPLE)
     list(APPEND memory_tools_extra_test_env
       DYLD_INSERT_LIBRARIES=$<TARGET_FILE:memory_tools_interpose>)
   elseif(UNIX)
-    list(APPEND memory_tools_extra_test_env
-      LD_PRELOAD=$<TARGET_FILE:memory_tools_interpose>)
+    set(memory_tools_path $<TARGET_FILE:memory_tools_interpose>)
+    if(OPTIONAL_ASAN_LD_PRELOAD_OVERRIDE)
+      execute_process(COMMAND bash "-c" "ldconfig -p | grep -e "libasan.*x86_64" | awk '{print $NF}'"
```

Comment:
This is an improvement over an `ls`-based filesystem approach but still not foolproof. I would really like to find a way to get the same shared object the linker would.

Reply:
Let me explore if there is anything else we can use here and get back on this :)

Reply:
Ideally, the method would also work on macOS (as clang is the default compiler there).

Reply:
The other approach I am able to find online is to perform `ldd` on an executable. So it will be on the lines of running `ldd` on the "test_runner" executable inside the build directory for osrf_testing_tools. Trying to see a good way to do this or any other alternative way.

@wjwwood, I limited the above logic to linux mainly because ldconfig is present only on linux

Reply:
The `ldd` approach will be safer for systems which have more than one copy of clang/asan installed.

You could use `otool -L` on macOS instead of `ldd`.

Reply:
You could also enable leak detection on macOS:

https://clang.llvm.org/docs/AddressSanitizer.html#memory-leak-detection

Reply:
Makes sense. I am trying to check if this can be used to fetch the path for libasan
$<TARGET_SONAME_FILE:tgt>
File with soname (.so.3) where tgt is the name of a target.
https://cmake.org/cmake/help/v3.0/manual/cmake-generator-expressions.7.html

Reply:
IMHO the best way is to use `-print-file-name` from GCC which can be used to get the location of GCC controlled files.
My understanding is that this option is recognized by both clang and GCC.
It's use e.g. here: https://github.com/gimli-org/gimli/blob/master/cmake/FindCHOLMOD.cmake to find gfortran.

Ultimately, what we want is to keep the compiler instrumentation in sync with the supporting library. To achieve that, querying the compiler is probably the safest path forward.

Reply:
@thomas-moulard that sounds ideal, but I'm not sure how it is supposed to work, on my macOS laptop it just seems to always return what you give it:

```
william@kisra ~/ros2_ws
% clang++ -print-file-name=libasan
libasan

william@kisra ~/ros2_ws
% clang++ -print-file-name=libasan.so
libasan.so

william@kisra ~/ros2_ws
% clang++ -print-file-name=libasan.dylib
libasan.dylib

william@kisra ~/ros2_ws
% gcc -print-file-name=libasan.dylib
libasan.dylib

william@kisra ~/ros2_ws
% gcc -print-file-name=libasan.so
libasan.so

william@kisra ~/ros2_ws
% gcc -print-file-name=libasan.so.3
libasan.so.3

william@kisra ~/ros2_ws
% gcc -print-file-name=libgfortran.so
libgfortran.so

william@kisra ~/ros2_ws
% gcc -print-file-name=libgfortran.dylib
libgfortran.dylib
```

Reply:
+1 on @thomas-moulard 's suggestion. Will test the changes with `gcc --print-file-name` once and then update here

Reply:
If I use a newer gcc it works, but is kind of fragile:

```
william@kisra ~/ros2_ws
% /usr/local/Cellar/gcc/8.2.0/bin/gcc-8 -print-file-name=libasan.dylib
/usr/local/Cellar/gcc/8.2.0/lib/gcc/8/gcc/x86_64-apple-darwin18.2.0/8.2.0/../../../libasan.dylib

william@kisra ~/ros2_ws
% /usr/local/Cellar/gcc/8.2.0/bin/gcc-8 -print-file-name=libasan
libasan

william@kisra ~/ros2_ws
% /usr/local/Cellar/gcc/8.2.0/bin/gcc-8 -print-file-name=libasan.so
libasan.so
```

So use it with care.

Reply:
@wjwwood - Updated the PR with the above changes. I am still carrying out the testing for the same.